dashboard/app/routes/optimization/fine-tuning/route.tsx

```diff
@@ -0,0 +1,324 @@
+      console.error("Submission error:", error);
+    }
+  }
+
+  return (
+    <div className="min-h-screen bg-gray-50">
+      <main className="p-4">
+        <h2 className="scroll-m-20 border-b pb-2 text-3xl font-semibold tracking-tight first:mt-0">
+          Fine-Tuning
+        </h2>
+
+        <div className="mt-8">
+          <Form {...form}>
+            <form
+              onSubmit={(e) => {
+                e.preventDefault();
+
+                // Try calling onSubmit directly with current values
+                const values = form.getValues();
+                onSubmit(values);
+              }}
+              className="space-y-6"
+            >
+              <div className="space-y-6">
+                <FunctionSelector
+                  control={form.control}
+                  inferenceCount={counts.inferenceCount}
+                  config={config}
+                  onFunctionChange={handleFunctionChange}
+                />
+
+                <MetricSelector
+                  control={form.control}
+                  feedbackCount={counts.feedbackCount}
+                  curatedInferenceCount={counts.curatedInferenceCount}
+                  config={config}
+                  onMetricChange={handleMetricChange}
+                />
+
+                <VariantSelector
+                  control={form.control}
+                  chatCompletionVariants={getChatCompletionVariantsForFunction()}
+                />
+
+                <ModelSelector control={form.control} models={models} />
+
+                <AdvancedParametersAccordion control={form.control} />
+              </div>
+
+              <div className="space-y-4">
+                <Button
+                  type="submit"
+                  disabled={
+                    !form.watch("function") ||
+                    !form.watch("metric") ||
+                    !form.watch("model") ||
+                    !form.watch("variant") ||
+                    form.formState.isSubmitting ||
+                    isSubmitted
+                  }
+                >
+                  {getButtonText()}
+                </Button>
+
+                {loaderData && (
+                  <div className="p-4 bg-gray-100 rounded-lg mt-4">
+                    <div className="mb-2 font-medium">
+                      Loader Data (Last Updated:{" "}
+                      {new Date().toLocaleTimeString()})
+                    </div>
+                    <Textarea
+                      value={JSON.stringify(loaderData, null, 2)}
+                      className="w-full h-48 resize-none bg-transparent border-none focus:ring-0"
+                      readOnly
+                    />
+                  </div>
+                )}
+
+                {finalResult && (
+                  <div className="p-4 bg-gray-100 rounded-lg">
+                    <div className="mb-2 font-medium">Configuration</div>
+                    <Textarea
+                      value={finalResult}
```

Comment:
![CleanShot 2024-12-21 at 17 37 29@2x](https://github.com/user-attachments/assets/df87e1af-8410-4326-9730-dc599d6732b6)

Reply:
JavaScript `smol-toml` package does weird things, I added manual handling to clean this up and unit tested.